linalg/solvers.hpp

```diff
@@ -434,6 +465,43 @@ class NewtonSolver : public IterativeSolver
    /** @brief This method can be overloaded in derived classes to perform
        computations that need knowledge of the newest Newton state. */
    virtual void ProcessNewState(const Vector &x) const { }
+
+   const Vector &GetCurrentResidual() const { return r; }
+   const Vector &GetCurrentIterate() const { return xcur; }
+
+   /// Enable adaptive linear solver relative tolerance algorithm.
+   /** Compute a relative tolerance for the Krylov method after each nonlinear
+    iteration, based on the algorithm presented in [1].
+
+    The maximum linear solver relative tolerance @a rtol_max should be < 1.
+    For @a version 1 the parameters @a alpha and @a gamma are ignore. For @a
+    version 2 @a alpha has to be between 0 and 1 and @a alpha between 1 and 2.
+
+    [1] Eisenstat, Stanley C., and Homer F. Walker. "Choosing the forcing terms
+    in an inexact Newton method."
+    */
+   void SetAdaptiveLinRtol(const int version = 2,
+                           const double rtol0 = 0.5,
+                           const double rtol_max = 0.9,
+                           const double alpha = 0.5 * (1.0 + sqrt(5.0)),
+                           const double gamma = 1.0);
+
+   /** @brief Method for the adaptive linear solver rtol invoked before the
+       linear solve. */
+   void AdaptiveLinRtolPreSolve(const Vector &x,
+                                const int it,
+                                const double fnorm) const;
+
+   /** @brief Method for the adaptive linear solver rtol invoked after the
+       linear solve. */
+   void AdaptiveLinRtolPostSolve(const Vector &x,
+                                 const Vector &b,
+                                 const int it,
+                                 const double fnorm) const;
+
+   static void DQOpTimes(const Vector &v, Vector &Jv, void *ctx);
```

Comment:
These function pointers seem too intrusive (or ignore) the existing linear solvers' interface. Why don't you create a custom `Operator` that knows about the `NewtonSolver` and give it to the linear solver? This would avoid changes in `IterativeSolver`. For example, the `NewtonSolver` can pass this operator [here](https://github.com/mfem/mfem/pull/1818/files#diff-3d145a6c87b1ab35fc52ca1fe9793356cbde85403f3b872b0c7eca7c1a554a6eR1622), if approximate Hessians are enabled.

Reply:
My intention here was to circumvent the linear solver interface, otherwise we can't keep the preconditioning operators from the application perspective. If I overwrite the operator before the linear solve I'd have to memorize the original operator, no?

In a discussion with @v-dobrev we concluded that it might be a good addition to have a "callback" that resembles a custom Jacobian-Vector-Product in the IterativeSolver interface.

I'm all for a simpler, or more clear, approach, but I don't understand the proposed alternative.

Reply:
I must be missing something. I'll discuss in the context of `NewtonSolver` as it is the main interest of the PR.

You have an approximate `Mult` (which doesn't need the matrix assembly) for the gradient operator , and you give it to the linear solver, and the linear solver uses it. But then you always compute the original operator (assemble the matrix), as a preconditioner might need it.

Then what is the point of doing the approximate `Mult` if the exact action is available / computed anyway?

Reply:
Ok. If the exact action is available, it could definitely be used. However, in practice it shows often disadvantages since the "correctness" (or rather incorrectness) of the gradient and the relative tolerance algorithms can be out of agreement. It could be made optional to use DQ.